metadata/ota_metadata/metadata_gen.py

```diff
@@ -191,15 +191,59 @@ def gen_metadata(
     target_abs = Path(os.path.abspath(target_dir))
     ignore = ignore_rules(target_dir, ignore_file)
 
+    # If ignore file has the following directories:
+    #    1, "home/autoware/*/build"
+    #    2, "home/autoware/*/src"
+    # We will NOT simply ignore files under them.
+    # We will check the following:
+    #    1, If the file is set as the target of a symblink ?
+    #    2, If the file falls in some special folder pattern ?
+    # Why we need to do this? It is explained in this document
+    # https://tier4.atlassian.net/wiki/x/JoC21Q
+    check_patterns = [
+        re.compile(r"home/autoware/[^/]+/build"),
+        re.compile(r"home/autoware/[^/]+/src"),
+    ]
+
+    # This is the flag to control if we will check and add files back to "build" and "src" folder
+    check_symlink = any(
+        _pattern.search(str(_rule))
+        for _rule in ignore.rules
+        for _pattern in check_patterns
+    )
+
+    # Special Patterns that we need to check and add files.
+    build_folder_patterns = [
+        re.compile(r"home/autoware/[^/]*/build/.*/hook/.*"),
+        re.compile(r"home/autoware/[^/]*/build/.*/.*.egg-info/.*"),
+    ]
+
+    additional_symlink_set = set()
+    additional_regular_set = set()
+    additional_dir_set = set()
+
     # remove kernels under /boot directory other than latest
     non_latest_kernels = _list_non_latest_kernels(p / "boot")
-
     dirs = []
     symlinks = []
     regulars = []
     for f in p.glob("**/*"):
         try:
             if ignore.match(target_abs / str(f.relative_to(target_dir))):
+                if check_symlink:
+                    relative_path = str(f.relative_to(target_dir))
+                    if any(
+                        re.search(pattern, relative_path) for pattern in check_patterns
+                    ):
+                        if f.is_symlink():
+                            additional_symlink_set.add(relative_path)
+                        elif f.is_dir():
+                            additional_dir_set.add(relative_path)
+                        elif f.is_file() and any(
+                            _file_pattern.search(relative_path)
+                            for _file_pattern in build_folder_patterns
+                        ):
+                            additional_regular_set.add(relative_path)
```

Comment:
Oh that is the reason why you still add all the dirs and symlinks under `build/` and `src/` folders, this can handle nested symlinks and parents dirs case here.